### PR TITLE
[RFR] Add Dockerized database dump on S3 post

### PR DESCRIPTION
**Pull request by jpetitcolas**
image

### CODE REVIEW

**Comment by djhi** [comment]
Nice one :+1:, I haven't noticed any error

File: blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+image from Docker Hub setting these variables at container creation.
+
+We then copy the file from the container to the host using the `docker cp`
+command, and clean the container temporary folder.
+
+## Keeping only last backups
+
+We generally keep data backups for a week. We then have 7 days to detect some data
+anomalies and to restore a dump. Some developers prefers to keep another backup
+per month, but except in very sensitive environment, data are too outdated to be
```
Comment (fzaninotto) [inline blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md]: s/environment/environments

Comment (fzaninotto) [inline blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md]: s/data are too outdated/month-old data is too old/

File: blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+useful. So, let's focus on last 7 days:
```
Comment (fzaninotto) [inline blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md]: s/on last/on the last

File: blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+one, process would be quite the same, just replace `pg_dump` by `mysqldump`. We
+put the dump into the container `/tmp` folder.
+
+Note the `sh -c` command. This way, we can pass a whole command (including file
+redirections) as a string, without worrying about conflicts between host and
+container paths.
+
+Database credentials are passed via environment variables. Note the `\` inside
+the Docker shell command: it is important as we want these variables to be
+interpreted in the Docker container, not when host interpret the command. In
```
Comment (fzaninotto) [inline blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md]: s/not when host interpret/not when the host interprets

File: blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+this example, we're using the [official PostGreSQL](https://hub.docker.com/_/postgres/)
+image from Docker Hub setting these variables at container creation.
```
Comment (fzaninotto) [inline blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md]: s/Docker Hub setting/Docker Hub, setting

File: blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+image: /images/blog/cloud-backup.png
+author: <a href="http://www.jonathan-petitcolas.com">Jonathan Petitcolas</a>
+canonical: http://www.jonathan-petitcolas.com/2015/09/28/dump-docker-ized-database-to-amazon-s3.html
+tags:
+- Docker
+- Amazon S3
+- database
+---
+
+We sometimes use Docker for our staging environments. Who says staging or production
```
Comment (fzaninotto) [inline blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md]: "who says" is frenglish. Find another more idiomatic sentence.

File: blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+also says data backup. Here is a post describing how we regularly upload our database
```
Comment (fzaninotto) [inline blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md]: supload our database/upload database

File: blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+dumps directly from a Docker container to Amazon S3 servers.
+
+## Dumping Docker-ized database on host
+
+First step is to create a dump of our database. Let's consider we have an
```
Comment (fzaninotto) [inline blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md]: s/First step/The first step

Comment (fzaninotto) [inline blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md]: s/a dump of our database/a dump of the database

File: blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+
+# Backup from docker container
+docker exec $CONTAINER_NAME sh -c "PGPASSWORD=\"\$POSTGRES_PASSWORD\" pg_dump --username=\$POSTGRES_USER \$POSTGRES_USER > /tmp/$FILENAME"
+docker cp $CONTAINER_NAME:/tmp/$FILENAME $DUMPS_FOLDER
+docker exec $CONTAINER_NAME sh -c "rm /tmp/awesomeproject-*.sql"
+```
+
+After setting a few configuration variables to ease re-use of this script, we
+execute via the `docker exec` command a dump of our PSQL database. For a MySQL
+one, process would be quite the same, just replace `pg_dump` by `mysqldump`. We
```
Comment (fzaninotto) [inline blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md]: s/process/the process

Comment (fzaninotto) [inline blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md]: s/quite the same, just replace/quite the same -- just replace

File: blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+per month, but except in very sensitive environment, data are too outdated to be
+useful. So, let's focus on last 7 days:
+
+``` sh
+# Keep only 7 most recent backups
+cd $DUMPS_FOLDER && (ls -t | head -n 7 ; ls -t) | uniq -u | xargs --no-run-if-empty rm
+cd $DUMPS_FOLDER && bzip2 --best $FILENAME
+```
+
+The first command probably looks like voodoo. Let's explicit it. First step, we
```
Comment (fzaninotto) [inline blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md]: s/Let's explicit it/Let's explain it

Comment (fzaninotto) [inline blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md]: s/First step/First

File: blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+go into the dumps folder. Note we need to repeat it on all following lines.
```
Comment (fzaninotto) [inline blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md]: s/Note we need/Note that we need

File: blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+Indeed, commands are executed each in their own sub-process. Thus, the `cd` only
+affect current command, on a single line.
```
Comment (fzaninotto) [inline blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md]: s/affect current command/affects the current command

File: blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md [hunk 12]
```diff
@@ -0,0 +1,254 @@
...
+The `(ls -t | head -n 7 ; ls -t)` command is a parallel command. We actually
+execute two different commands at the same time, to send their result on the
+standard output. We list (`ls`) our files by modification time (`-t`), newer
+first. We then keep the 7 first lines (`head -n 7`).
+
+In addition, we also display all available files. So, all files we should keep
+are displayed twice.
+
+Finally, we keep only the unique lines, using `uniq -u`, and `rm` each result.
+The `--no-run-if-empty` is just to prevent an error when no file needs to be
```
Comment (fzaninotto) [inline blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md]: s/The `--no-run-if-empty`/The `--no-run-if-empty` option

File: blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md [hunk 13]
```diff
@@ -0,0 +1,254 @@
...
+- Docker
+- Amazon S3
+- database
+---
+
+We sometimes use Docker for our staging environments. Who says staging or production
+also says data backup. Here is a post describing how we regularly upload our database
+dumps directly from a Docker container to Amazon S3 servers.
+
+## Dumping Docker-ized database on host
```
Comment (fzaninotto) [inline blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md]: Use all-caps for section tiles

File: blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md [hunk 14]
```diff
@@ -0,0 +1,254 @@
...
+
+Finally, we keep only the unique lines, using `uniq -u`, and `rm` each result.
+The `--no-run-if-empty` is just to prevent an error when no file needs to be
+deleted.
+
+As we pay Amazon depending of the amount of data stored on S3, we also
+compress at the maximum our dumps, using `bzip2 --best`. Even if prices are
+really cheap, it is always worthy to use a single command to save money, isn't it?
+
+## Uploading database dumps on Amazon S3
```
Comment (fzaninotto) [inline blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md]: s/on/to

File: blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md [hunk 15]
```diff
@@ -0,0 +1,254 @@
...
+
+Storing dumps on the same machine than databases is not a good idea. Hard disk
```
Comment (fzaninotto) [inline blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md]: s/Hard disk/Hard disks

File: blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md [hunk 16]
```diff
@@ -0,0 +1,254 @@
...
+may crash, user may block itself from connecting because of too restrictive
```
Comment (fzaninotto) [inline blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md]: s/user may block itself/users may block themselves

File: blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md [hunk 17]
```diff
@@ -0,0 +1,254 @@
...
+Unlimited FTP Servers", according to the [AWS in Plain English](https://www.expeditedssl.com/aws-in-plain-english)
+post. It is used to:
+
+> Store images and other assets for websites. Keep backups and share files
+between services. Host static websites. Also, many of the other AWS services
+write and read from S3.
+
+Of course, you can restrict access to your files, as we are going to configure it.
+
+### Creating bucket
```
Comment (fzaninotto) [inline blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md]: s/Creating bucket/Creating A Bucket

File: blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md [hunk 18]
```diff
@@ -0,0 +1,254 @@
...
+
+First step is to create a S3 bucket. You can replace `bucket` by `hosting` for a
+better understanding. So, log on your [Amazon Web Services (AWS) console](https://console.aws.amazon.com/console/home),
+and select S3. Then, `Create bucket` and give it a name and locate it.
```
Comment (fzaninotto) [inline blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md]: s/`Create bucket` and give it a name and locate it/`Create bucket`, give it a name, and choose a location for it

File: blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md [hunk 19]
```diff
@@ -0,0 +1,254 @@
...
+
+<div class="image">
+    <img src="/images/blog/db-dumps/s3-create-bucket.png" alt="Create a bucket on Amazon S3" title="Create a bucket on Amazon S3" class="responsive" />
+</div>
+
+Keep defaut values for all others parameters. We are now going to restrict access
+to our bucket creating a specific user.
```
Comment (fzaninotto) [inline blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md]: s/bucket creating/bucket, by creating

File: blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md [hunk 20]
```diff
@@ -0,0 +1,254 @@
...
+
+### Creating restricted access to our bucket
```
Comment (fzaninotto) [inline blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md]: s/Creating restricted access to our bucket/Restricting Access To The Bucket

File: blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md [hunk 21]
```diff
@@ -0,0 +1,254 @@
...
+
+This is the only painful step: creating a user with correct permissions to
+access our bucket. Indeed, our server script needs to connect on AWS. We can of
```
Comment (fzaninotto) [inline blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md]: s/to connect on AWS/to connect to AWS

File: blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md [hunk 22]
```diff
@@ -0,0 +1,254 @@
...
+                "arn:aws:s3:::awesomeproject-private/*"
+            ]
+        }
+    ]
+}
+```
+
+This policy allows three operations, respectively:
+
+* Listing all buckets available to user (required to connect on S3),
```
Comment (fzaninotto) [inline blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md]: s/Listing/List

File: blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md [hunk 23]
```diff
@@ -0,0 +1,254 @@
...
+* List bucket content on the specified resource,
+* Write, read and delete permissions on bucket content.
+
+<p class="image">
+    <img src="/images/blog/db-dumps/s3-policy.png" alt="Create a Amazon IAM policy" title="Create a Amazon IAM policy" class="responsive" />
+</p>
+
+Validate the policy and create a new user (in the `Users` menu). Do not forget to
+save her credentials, as we will never be able to retrieve our secret access key
```
Comment (fzaninotto) [inline blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md]: s/her/their

File: blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md [hunk 24]
```diff
@@ -0,0 +1,254 @@
...
+
+Finally, let's cron our script to automatically save our data each day. Add the
+following file into `/etc/cron.d/awesomeproject`:
+
+```
+# Daily DB backup
+00 23 * * * awesomeproject /home/ubuntu/awesomeproject/bin/db-save.sh >> /home/ubuntu/cron.out.log 2>&1
+```
+
+It launches the script every day at 11pm as `awesomeproject` user. Rather use
```
Comment (fzaninotto) [inline blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md]: s/Rather use absolute paths instead of relative one/Prefer absolute paths to relative ones

File: blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md [hunk 25]
```diff
@@ -0,0 +1,254 @@
...
+absolute paths instead of relative one. It would prevent you from serious headaches.
```
Comment (fzaninotto) [inline blog/_posts/2015-10-26-dump-docker-ized-database-to-amazon-s3.md]: s/It would prevent you from serious headaches./It will save you some serious headaches

**Comment by fzaninotto** [comment]
Would you mind changing the publication date to tomorrow?

**Comment by jpetitcolas** [comment]
Code review applied. Ready for publication, today.
